src/util/newconfig/ConfigDefinition.cpp

```diff
@@ -0,0 +1,213 @@
+     {"io_threads", ConfigValue{ConfigType::Integer}.defaultValue(2)},
+     {"cache.num_diffs", ConfigValue{ConfigType::Integer}.defaultValue(32)},
+     {"cache.num_markers", ConfigValue{ConfigType::Integer}.defaultValue(48)},
+     {"cache.page_fetch_size", ConfigValue{ConfigType::Integer}.defaultValue(512)},
+     {"cache.load", ConfigValue{ConfigType::String}.defaultValue("async")},
+     {"log_channels.[].channel", Array{ConfigValue{ConfigType::String}.optional()}},
+     {"log_channels.[].log_level", Array{ConfigValue{ConfigType::String}.optional()}},
+     {"log_level", ConfigValue{ConfigType::String}.defaultValue("info")},
+     {"log_format",
+      ConfigValue{ConfigType::String}.defaultValue(
+          R"(%TimeStamp% (%SourceLocation%) [%ThreadID%] %Channel%:%Severity% %Message%)"
+      )},
+     {"log_to_console", ConfigValue{ConfigType::Boolean}.defaultValue(false)},
+     {"log_directory", ConfigValue{ConfigType::String}.optional()},
+     {"log_rotation_size", ConfigValue{ConfigType::Integer}.defaultValue(2048)},
+     {"log_directory_max_size", ConfigValue{ConfigType::Integer}.defaultValue(50 * 1024)},
+     {"log_rotation_hour_interval", ConfigValue{ConfigType::Integer}.defaultValue(12)},
+     {"log_tag_style", ConfigValue{ConfigType::String}.defaultValue("uint")},
+     {"extractor_threads", ConfigValue{ConfigType::Integer}.defaultValue(2)},
+     {"read_only", ConfigValue{ConfigType::Boolean}.defaultValue(false)},
+     {"txn_threshold", ConfigValue{ConfigType::Integer}.defaultValue(0)},
+     {"start_sequence", ConfigValue{ConfigType::String}.optional()},
+     {"finish_sequence", ConfigValue{ConfigType::String}.optional()},
+     {"ssl_cert_file", ConfigValue{ConfigType::String}.optional()},
+     {"ssl_key_file", ConfigValue{ConfigType::String}.optional()},
+     {"api_version.min", ConfigValue{ConfigType::Integer}},
+     {"api_version.max", ConfigValue{ConfigType::Integer}}}
+};
+
+ObjectView
+ClioConfigDefinition::getObject(std::string_view prefix, std::optional<std::size_t> idx) const
+{
+    std::string prefixWithDot = std::string(prefix) + ".";
+    for (auto const& [mapKey, mapVal] : map_) {
+        if (mapKey.starts_with(prefixWithDot))
+            ASSERT(!mapKey.ends_with(".[]"), "Trying to retrieve an object when value is an Array");
+
+        if (mapKey.starts_with(prefixWithDot) && std::holds_alternative<ConfigValue>(mapVal))
+            return ObjectView{prefix, *this};
+
+        if (mapKey.starts_with(prefixWithDot) && std::holds_alternative<Array>(mapVal)) {
+            ASSERT(std::get<Array>(mapVal).size() > idx, "index provided is out of scope");
+            return ObjectView{prefix, idx.value(), *this};
+        }
+    }
+    throw std::invalid_argument("Key is not found in config");
+}
+
+ArrayView
+ClioConfigDefinition::getArray(std::string_view prefix) const
+{
+    std::string key = std::string(prefix);
+    if (!prefix.contains(".[]"))
```

Comment:
nit: do we support ends_with?

Reply:
yeah, this could be just `prefix.ends_with(..)`, even without the temporary. We already use string's .ends_with in the codebase. I think string_view support should be on par with string but who knows - need to try.

Reply:
Shouldn't be .ends_with() because we can get ArrayView by doing getArray(etl_sources.[].ip) 😅 👍